setup.py

```diff
@@ -101,6 +101,16 @@ def compilation_help():
     os.environ['CFLAGS'] = cflags
     sys.argv.remove ('-pygame-ci')
 
+# This should detect Raspberry pis at least.
+from platform import machine
+if machine().startswith('arm'):
+    cflags = os.environ.get('CFLAGS', '')
+    if cflags:
+        cflags += ' '
+    cflags += '-mfpu=neon'
+    os.environ['CFLAGS'] = cflags
```

Comment:
I think this will apply to all files, making all fpu code use neon - even outside this blit routine.

For binary wheels, built on say a Debian or piwheels build server, they would compile with neon. But then a computer on say a RPI 1 loading that wheel would fail (I think).

Reply:
Yes, I think you are right. 

I also don't think that you can upload wheels that are that specific to PyPI, it looks like the 2014 'manylinux' standard sets armv7 as minimum suggesting that there is no support for armv6 differentiation.

I guess the option that SDL, openCV and tensorflow are using is to upload wheels/binaries  _without_ neon optimisations and then allowing a build flag to enable them when you build from source so users can choose to build with it or not. I think we should probably do the same.

SDL link:
https://www.phoronix.com/scan.php?page=news_item&px=SDL-ARM-NEON-Optimizations

OpenCV link for ENABLE_NEON flag:
https://www.pyimagesearch.com/2017/10/09/optimizing-opencv-on-the-raspberry-pi/

Reply:
Looking at SDL's CMake it appears they set the ARMNEON option _on_ by default here:
https://hg.libsdl.org/SDL/file/4298bf108b06/CMakeLists.txt#l332

Which then tries neon fpu in this block here I believe:
https://hg.libsdl.org/SDL/file/4298bf108b06/CMakeLists.txt#l704

Which does a little check to see if it can compile some assembly source code with ARM blitters before actually enabling them. So perhaps we can borrow this `check_c_source_compiles` check from SDL and use that to toggle the fpu flag on or off?

Reply:
Note, that raspberian is set up to load modules from https://www.piwheels.org/ first (see /etc/pip.conf). They try and keep everything compatible with `armv6`, because they still ship new hardware with that (the zero's), and this is the reason for the Raspberian distro - since Debian supported a higher version. https://www.raspbian.org/RaspbianFAQ#What_is_Raspbian.3F

It's disappointing that the manylinux effort messed up with raspberian support. At least now there is i386 and some ARM support in 2014. I haven't kept up with it the changes in the last half year.

---

> I guess the option that SDL, openCV and tensorflow are using is to upload wheels/binaries without neon optimisations and then allowing a build flag to enable them when you build from source so users can choose to build with it or not. I think we should probably do the same.

I guess compiling it with `CFLAGS=-mfpu=neon setup.py install` would work for this option? Platforms with these flags built in won't need to do that either.


---

Here's another way we could hack up distutils based on python modules and compilation flags. (obviously these ways would be out of scope for this PR)

Currently Setup.SDLX.in has stuff like this.
```
surface src_c/surface.c src_c/alphablit.c src_c/surface_fill.c $(SDL) $(DEBUG)
```

Adding CPU specific modules could be done like this:
```
surface_fpu_neon src_c/surface.c src_c/alphablit.c src_c/surface_fill.c $(SDL) $(DEBUG) -mfpu=neon
```
Then the init machinery (`__init__.py`) can load the correct `surface` (or `surface_fpu_neon`) module based on the CPU capabilities. I guess all the other C based routines would also be sped up because of better compiler optimizations too.

This is obviously complicated. I feel if we only rely on compiler flags, then most users will not be able to take advantage of the optimizations -- because most users will use the pre-built binaries.

Reply:
So, the way forward for this PR is to remove the setup.py changes, and to require people to add the CFLAGS themselves on build?

Reply:
What I added in the latest commit was a flag for setup.py called `-ENABLE_NEON` mainly so that I could also add a matching define (PG_ENABLE_ARMNEON) to the c preprocessor so that the compiler instructions match whats in the code. 

In theory you should be able to enable the ARM intrinsics like this:

`python3 setup.py install -ENABLE_NEON`

This was me mimicking what is done for the CI build configuration which has this flag:

`python3 setup.py install -pygame-ci`

I'm not attached to the all-caps, probably they should be consistent.

Previously, I was using the SDL defines for including and removing the ARM code - but I realised there could be a mismatch between whatever SDL was doing with ARM and what pygame was enabling if we didn't have our own define.

I still need to run tests on my Pi4 with and without the flag to check it works but I think it's a reasonable stop gap for this pull request.

------

I think your idea of pulling the ARM code out into it's own .c file compiled with the flag for the wheels is a good one though. I suppose that since we have a runtime check for NEON capabilities as well we could just drag the actual function body over to the .c file compiled with `mfpu-neon` and import a header with just the function declarations. That the .c body wouldn't run on a Pi1 wouldn't matter because the Pi1 would never call the function in the first place.

Reply:
> So, the way forward for this PR is to remove the setup.py changes, and to require people to add the CFLAGS themselves on build?

I had a look at numpy and I _think_ they solve this problem by using function attributes.  But then, numpy seems to generate half their performance critical code from templates, so I am not sure it is trivial to apply directly to our situation.

Reply:
... continuing this discussion here: [CPU specific modules (neon, SSE4, etc)](https://github.com/pygame/pygame/issues/1723)